tendrl/gluster_integration/objects/volume/atoms/shrink/__init__.py

```diff
@@ -0,0 +1,144 @@
+
+
+class Shrink(objects.BaseAtom):
+    obj = Volume
+    def __init__(self, *args, **kwargs):
+        super(Shrink, self).__init__(*args, **kwargs)
+
+    def run(self):
+        args = {}
+        vol = Volume(vol_id=self.parameters['Volume.vol_id']).load()
+        if self.parameters.get('Volume.replica_count') is not None:
+            args.update({
+                "replica_count": self.parameters.get('Volume.replica_count')
+            })
+            if vol.replica_count != self.parameters.get('Volume.replica_count'):
+                args.update({"decrease_replica_count": True})
+        elif self.parameters.get('Volume.disperse_count') is not None:
+            args.update({
+                "disperse_count": self.parameters.get('Volume.disperse_count')
+            })
+        else:
+            if int(vol.replica_count) > 1:
+                args.update({
+                    "replica_count": vol.replica_count
+                })
+            elif int(vol.disperse_count) > 1:
+                args.update({
+                    "disperse_count": vol.disperse_count
+                })
+                
+        if self.parameters.get('Volume.force') is not None:
+            args.update({
+                "force": self.parameters.get('Volume.force')
+            })
+
+        action = self.parameters.get('Volume.action')
+
+        Event(
+            Message(
+                priority="info",
+                publisher=NS.publisher_id,
+                payload={
+                    "message": "Shrinking the volume %s" %
+                    self.parameters['Volume.volname']
+                },
+                job_id=self.parameters["job_id"],
+                flow_id=self.parameters["flow_id"],
+                cluster_id=NS.tendrl_context.integration_id,
+            )
+        )
+
+        if NS.gdeploy_plugin.shrink_volume(
+                self.parameters.get('Volume.volname'),
+                self.parameters.get('Volume.bricks'),
+                action,
+                **args
+        ):
+            Event(
+                Message(
+                    priority="info",
+                    publisher=NS.publisher_id,
+                    payload={
+                        "message": "Shrinked the volume %s" %
+                        self.parameters['Volume.volname']
+                    },
+                    job_id=self.parameters["job_id"],
+                    flow_id=self.parameters["flow_id"],
+                    cluster_id=NS.tendrl_context.integration_id,
+                )
+            )
+
+            if action != "commit" and not args.has_key("decrease_replica_count"):
+                return True
+
+            # Delete the bricks from central store
+            # Acquire lock before deleting the bricks from etcd
+            # We are blocking till we acquire the lock
+            # the lock will live for 60 sec after which it will be released.
+            lock = etcd.Lock(NS.etcd_orm.client, 'volume')
```

Comment:
Since you are only handling a specific volume ($volume.vol_id), it is better to use a volume specific lock here, instead of using the generic "volume" lock which locks up all the other volumes too